src/Orleans.Core/Transactions/ITransactionAgent.cs

```diff
@@ -22,16 +22,13 @@ public interface ITransactionAgent
         Task<ITransactionInfo> StartTransaction(bool readOnly, TimeSpan timeout);
 
         /// <summary>
-        /// Attempt to Commit a transaction. If this returns with no exceptions
-        /// then the transaction is successfully committed.
+        /// Attempt to Commit a transaction. 
         /// </summary>
         /// <param name="transactionInfo">transaction info</param>
-        /// <returns>None.</returns>
-        /// <remarks>
-        /// The method throws OrleansTransactionInDoubtException if the outcome of the Commit cannot be determined.
-        /// If any other exception is thrown then the transaction is aborted.
-        /// </remarks>
-        Task Commit(ITransactionInfo transactionInfo);
+        /// <returns>null if the transaction committed successfully, or an exception otherwise.
+        /// If the exception is OrleansTransactionInDoubtException, it means the outcome of the Commit cannot be determined; otherwise,
+        /// the transaction is guaranteed to not have taken effect.</returns>
+        Task<OrleansTransactionException> Commit(ITransactionInfo transactionInfo);
```

Comment:
Why have an exception as a return value, this seems very odd and unorthodox.  How is receiving an exception different from the call throwing an exception.

Reply:
Exceptions only capture their stack when they are thrown, so returning an exception avoids that.

I would very much prefer we return a result type (some struct or class) which represents the commit rather than an exception

Reply:
I will change this to return TransactionalStatus instead of an exception.